### PR TITLE
[prometheusremotewrite] add otel_scope_info metric to prometheus remote write exporter

### PR DESCRIPTION
**Pull request by blakeroberts-wk**
**Description:** <Describe what has changed.>

These changes add scope name and version to a metric data point as labels. Additionally, an `otel_scope_info` metric may be emitted if the scope contains additional attributes. Similar to the `target_info` metric, the `otel_scope_info` metric can be joined with a metric on `job`, `instance`, `otel_scope_name`, and `otel_scope_version` to retrieve instrumentation scope attributes.

These changes align the prometheus remote write exporter with the following specification update: https://github.com/open-telemetry/opentelemetry-specification/pull/2703

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21091

**Testing:** <Describe what testing was performed and which tests were added.>

I pulled these changes into a collector which accepts OTLP metrics from many java agents and exports data to a prometheus compatible backend.

```tf
// series 1
http_server_active_requests{
http_method="GET",
http_scheme="http",
instance="REDACTED_1",
job="REDACTED_2",
net_host_name="REDACTED_3",
net_host_port="8080",
otel_scope_name="io.opentelemetry.servlet-5.0",
otel_scope_version="1.26.0-alpha"
}
// series 2
http_server_active_requests{
http_method="GET",
http_scheme="http",
instance="REDACTED_1",
job="REDACTED_2",
net_host_name="REDACTED_3",
net_host_port="8080",
otel_scope_name="io.opentelemetry.tomcat-10.0",
otel_scope_version="1.26.0-alpha"
}
```

After deployment, errors from the backend around samples with duplicate timestamps ceased.

**Documentation:** <Describe the documentation added.>

The prometheus remote write exporter's README was updated with the new configuration setting to enable/disable the creation of the `otel_scope_info` metric. Like the `target_info` metric, I've enabled this feature by default which may or may not be appropriate.

### CODE REVIEW

File: exporter/prometheusremotewriteexporter/factory.go
Comment (Aneurysm9) [inline exporter/prometheusremotewriteexporter/factory.go:101]: I don't think this should default to `true`.  We can perhaps revisit this at a later date, but I would be more comfortable starting from a position that doesn't alter existing systems by default.

File: pkg/translator/prometheusremotewrite/helper.go
Comment (Aneurysm9) [inline pkg/translator/prometheusremotewrite/helper.go:220]: Should version be added as a label if name does not exist?  That implies bad data to me and I would omit both values.

File: pkg/translator/prometheusremotewrite/helper_test.go
Comment (Aneurysm9) [inline pkg/translator/prometheusremotewrite/helper_test.go:263]: There doesn't seem to be a test for a scope with name but without version, or the inverse.  My read of [the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#instrumentation-scope) is that name is required and version is not (though I'd give a strong side eye to anyone not including one).

File: exporter/prometheusremotewriteexporter/factory.go
Comment (blakeroberts-wk) [reply exporter/prometheusremotewriteexporter/factory.go:101]: I think this makes sense.

**Comment by blakeroberts-wk** [comment]
@evan-bradley would you review my changes, please? Thank you.

File: pkg/translator/prometheusremotewrite/helper.go
Comment (evan-bradley) [reply pkg/translator/prometheusremotewrite/helper.go:220]: I looked and most exporters in contrib will add the two independently, however I did find [one place](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/translator/loki/encode.go#L115) where `version` is excluded if `name` isn't present.

That said, I'm not clear on what the right solution is here. [Instrumentation Scope](https://opentelemetry.io/docs/specs/otel/glossary/#instrumentation-scope) is defined as identified through `name` or a `(name,version)` pair, so `version` on its own doesn't seem like it would be helpful. However, the guide for [mapping OTLP to non-OTLP](https://opentelemetry.io/docs/specs/otel/common/mapping-to-non-otlp/#instrumentationscope) doesn't mention that `name` is required for `version` to be present, and while it doesn't directly apply here, the [proto definitions say](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/common/v1/common.proto#L72) that an empty `name` field means it is unknown.

I'd probably lean toward saying we should just include the version, though I'm pretty conflicted.

**Comment by evan-bradley** [comment]
@blakeroberts-wk Sorry for the delay. Overall what you have looks good, but I don't know that I can answer the question of how we should handle emitting an Instrumentation Scope metric or name/version attributes if we don't know the name of the scope.

Please add a changelog entry for this change and fix the branch conflict.

File: pkg/translator/prometheusremotewrite/helper.go
Comment (dashpole) [reply pkg/translator/prometheusremotewrite/helper.go:228]: I'm also in-favor of including version if the name is not present.  Unless it is explicitly prohibited by the spec, we should be flexible with what we accept here.

File: pkg/translator/prometheusremotewrite/helper.go
Comment (dashpole) [inline pkg/translator/prometheusremotewrite/helper.go:607]: nit: scope info and target info are different concepts.  I'd prefer naming this addScopeInfo()

**Review by dashpole — APPROVED** [review]

File: .chloggen/O11Y-3152.yaml
Comment (evan-bradley) [inline .chloggen/O11Y-3152.yaml:12]:
```suggestion
note: Emit `otel_scope_info` metric and add `otel_scope_name` and `otel_scope_version` labels to metrics
```
I think we should also call out that the scope labels will be added to metrics. I did some wordsmithing here as I was trying to add that info, feel free to word this how you would like.

File: pkg/translator/prometheusremotewrite/helper.go
Comment (evan-bradley) [reply pkg/translator/prometheusremotewrite/helper.go:228]: The outcome from the 2023-08-02 NA/EU Collector SIG meeting discussion on this is that we would like to emit this data regardless of whether it is correct.

**Review by evan-bradley — APPROVED** [review]

**Comment by evan-bradley** [comment]
>@evan-bradley I have updated the branch and added a change log entry. are there any more concerns around these changes?

I don't have any concerns, but before we can merge a functional change like this we need approval from one of the code owners.

@Aneurysm9 @rapphil Please take a look when you are able.

File: pkg/translator/prometheusremotewrite/helper.go
Comment (rapphil) [inline pkg/translator/prometheusremotewrite/helper.go:618]: > It will contain the name and version of the instrumentation scope. Isn't this information per se useful?

It will contain the name and version of the instrumentation scope. Isn't this information per se useful? It means that there is at least a single metric being emitted in that scope.

Having `scope_info` in the configuration enabled and not generating the `otel_scope_info` metric is misleading and will confuse users.

File: pkg/translator/prometheusremotewrite/metrics_to_prw.go
Comment (rapphil) [inline pkg/translator/prometheusremotewrite/metrics_to_prw.go:111]: nit, make the signature respect the hierarchy: `addScopeTargetInfo(resource, scope, settings, mostRecentTimestamp, tsMap)`

File: pkg/translator/prometheusremotewrite/helper.go
Comment (blakeroberts-wk) [reply pkg/translator/prometheusremotewrite/helper.go:618]: > Isn't this information per se useful? It means that there is at least a single metric being emitted in that scope.

If one was looking for metrics under a specific scope, they'd probably do something like a topk query on
```
__name__{otel_scope_name="w/e"}
``` 
In general, actual usage of `otel_scope_info` is most likely after knowing the `__name__` you're after then
```
<query> / on (otel_scope_name, otel_scope_version) group_left otel_scope_info
```
Though the query would return no data if `otel_scope_info` doesn't exist. So if the feature is enabled, it might make sense to always report it so queries that use the metric always return data regardless of whether or not the joined labels add novel information

Comment (blakeroberts-wk) [reply pkg/translator/prometheusremotewrite/helper.go]: I pushed a commit that emitted the metric regardless.

But now I'm second guessing it. This has the potential to add a lot of active time series that won't be valuable the majority of the time. The collector could receive metrics from many languages and only PHP and rust even have the potential to make this metric add novel information. And the query with the join tends to be a rather advanced use case (so advanced that documentation doesn't exist around how to properly use `otel_scope_info` let alone `target_info`) Additionally, not emitting the metric if it doesn't have attributes would align with the implementation of `target_info`: https://github.com/blakeroberts-wk/opentelemetry-collector-contrib/blob/3f146d90bd9c622c33359c71fcae89162a263f2c/pkg/translator/prometheusremotewrite/helper.go#L589-L591

**Comment by blakeroberts-wk** [comment]
I'd like to address https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24248#discussion_r1293793538 by allowing the presence of the `otel_scope_info` time series when no scope attributes are present to be configurable.

In other words, along with the `enable` option, there would be an `alwaysEmit` option (or similarly named). It'd be set to true by default. But when explicitly set to false, it will avoid producing the `otel_scope_info` time series if the scope contained no additional attributes.

File: pkg/translator/prometheusremotewrite/helper.go
Comment (rapphil) [inline pkg/translator/prometheusremotewrite/helper.go:633]:
```suggestion
	labels := createAttributes(resCopy, scope, scope.Attributes(), settings.ExternalLabels, nameStr, name)
```

Comment (rapphil) [reply pkg/translator/prometheusremotewrite/helper.go:633]: However I think removing the namespace from the resource attributes is incorrect. 

According [to the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.25.0/specification/compatibility/prometheus_and_openmetrics.md#resource-attributes-1) namespace should be combined with the service name to form the job label.

>In OTLP, the service.name, service.namespace, and service.instance.id triplet is [required to be unique](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.25.0/specification/resource/semantic_conventions/README.md#service), which makes them good candidates to use to construct job and instance. In the collector Prometheus exporters, the service.name and service.namespace attributes MUST be combined as <service.namespace>/<service.name>, or <service.name> if namespace is empty,

`createAttributes` is already only considering the service.name, service.namespace, and service.instance.id triplet for resource attributes, therefore there is no need to create `resCopy`?

Comment (blakeroberts-wk) [reply pkg/translator/prometheusremotewrite/helper.go:638]: It would seem you are correct that `resCopy` is not necessary. I have removed its declaration.

File: pkg/translator/prometheusremotewrite/helper_test.go
Comment (blakeroberts-wk) [reply pkg/translator/prometheusremotewrite/helper_test.go:748]: I have added name and version to this test. Let me know if you'd like to see any additional variations.

**Review by dashpole — APPROVED** [review]

**Comment by blakeroberts-wk** [comment]
Any reservations for these changes?

@Aneurysm9 @evan-bradley @jpkrohling @rapphil

**Comment by evan-bradley** [comment]
@blakeroberts-wk Looks like some of the tests are failing, can you try rebasing/merging in main?

@Aneurysm9 @rapphil could you take a look as the code owners?

**Comment by blakeroberts-wk** [comment]
I have merged main.

The unit test `TestDeleteAfterRead_SkipPartials` within `github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer` passes when ran locally.

Can the CI check be re-ran or skipped? @evan-bradley

**Comment by blakeroberts-wk** [comment]
@evan-bradley are we still feeling confident in these changes?

**Comment by evan-bradley** [comment]
I don't see any issues with these changes, but we need an approval from the code owners of this component.

@Aneurysm9 @rapphil could you please take a look?

**Comment by evan-bradley** [comment]
@Aneurysm9 @rapphil Please take a look.

**Comment by Aneurysm9** [comment]
I think this is missing allowance for the use case where metrics are ingested with the `prometheus` receiver and expected to be exported without modification.  This is similar to the [discussions](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21743#issuecomment-1561727218) previously regarding the normalization of metric names that do not necessarily make sense when receiving and exporting prometheus data.  Configuration allowing the operator to control this behavior should be added.

**Comment by blakeroberts-wk** [comment]
> I think this is missing allowance for the use case where metrics are ingested with the `prometheus` receiver and expected to be exported without modification.

This implementation will not add the labels if the scope name/version are empty. However, if the scope name/version are empty, the prometheus receiver adds a default name and version: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9e47c92dc2cb7c4cf2f62f8c5afcd1724f520cd8/receiver/prometheusreceiver/internal/transaction.go#L233-L234.

I think it makes sense for the receiver to modify this data. But it does make it awkward to meet the expectation that prometheus data received by the collector isn't modified when exported in prometheus format. To be honest, I find this expectation to be silly. If I wanted to proxy prometheus data without modification, I'd use a prometheus server. If I wanted to proxy prometheus data with modification (which I do in practice), I'd use the otel collector. So to use the otel collector with the intent to modify prometheus data, I'd argue it'd be unreasonable for me to expect that data won't be modified. To be clear, modifications to the metric name seem intrusive. But adding labels used to appropriately identify a unique time series seems like reasonable, if not expected, modification.

Furthermore, the design docs of the prometheus receiver make the aforementioned expectation impossible:

> It doesn't need to provide one-to-one mapping
>
> https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/prometheusreceiver/DESIGN.md#map-prometheus-metrics-to-the-corresponding-opentelemetry-metrics-properly

Because the prometheus receiver does not guarantee data is mapped one-to-one, the prometheus remote write exporter cannot guarantee its output matches whatever input was given to the receiver.

> Configuration allowing the operator to control this behavior should be added.

This seems reasonable, but low value. Would it be acceptable to open an issue about adding this configuration?

**Comment by gespi1** [comment]
@evan-bradley @blakeroberts-wk what is the status of these changes, are they being superseded by another PR elsewhere or traction has stopped and still waiting on @Aneurysm9 ?

**Comment by evan-bradley** [comment]
I'm sorry, I lost track of this. This change was still dependent on the code owners' approval. I'm not aware of any other PRs that have superseded it.

**Comment by blakeroberts-wk** [comment]
@aglarendil I've been using a fork with these changes to handle related errors. I have not had the bandwidth to champion these changes through to completion. Feel free to take ownership.

**Pull request closed without merging**
